runtime/src/bank/check_transactions.rs

```diff
@@ -102,27 +110,46 @@ impl Bank {
                 nonce: None,
                 lamports_per_signature: hash_info.lamports_per_signature(),
             })
-        } else if let Some((nonce, nonce_data)) =
-            self.check_and_load_message_nonce_account(tx.message(), next_durable_nonce)
+        } else if let Some((nonce, lamports_per_signature)) =
+            self.check_load_and_advance_message_nonce_account(tx.message(), next_durable_nonce)
         {
             Ok(CheckedTransactionDetails {
                 nonce: Some(nonce),
-                lamports_per_signature: nonce_data.get_lamports_per_signature(),
+                lamports_per_signature,
             })
         } else {
             error_counters.blockhash_not_found += 1;
             Err(TransactionError::BlockhashNotFound)
         }
     }
 
-    pub(super) fn check_and_load_message_nonce_account(
+    pub(super) fn check_load_and_advance_message_nonce_account(
         &self,
         message: &SanitizedMessage,
         next_durable_nonce: &DurableNonce,
-    ) -> Option<(NonceInfo, nonce::state::Data)> {
+    ) -> Option<(NonceInfo, u64)> {
         let nonce_is_advanceable = message.recent_blockhash() != next_durable_nonce.as_hash();
         if nonce_is_advanceable {
-            self.load_message_nonce_account(message)
+            if let Some((nonce_address, mut nonce_account, nonce_data)) =
+                self.load_message_nonce_account(message)
+            {
+                let lamports_per_signature = nonce_data.get_lamports_per_signature();
+                let next_nonce_state = NonceState::new_initialized(
```

Comment:
This optimization can come later if we want it but I think we can delay the serialization of the advanced nonce by including the advanced `NonceData` in the `NonceInfo` and only serialize that data into the account shared data when `NonceInfo::account` is called. I think this would make the same fee payer / nonce account case a little cleaner too because we wouldn't need to copy the nonce account's data over during rollback accounts creation, we could delay it until `nonce.account()` is called when we collect accounts for storage.

Reply:
im going to leave this as is for now but may end up doing it in #2702 because ill need to reverify nonces mid-batch

Reply:
Sounds good to me. And just in case it wasn't fully clear, delaying the serialization will be nice because if the transaction succeeds, we won't need to waste any time serializing a rollback nonce account that is never used.